tombstone_gc.cc

```diff
@@ -0,0 +1,164 @@
+/*
+ * Copyright (C) 2021-present ScyllaDB
+ */
+
+/*
+ * This file is part of Scylla.
+ *
+ * Scylla is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Scylla is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Scylla.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <chrono>
+#include <boost/icl/interval.hpp>
+#include <boost/icl/interval_map.hpp>
+#include "schema.hh"
+#include "dht/i_partitioner.hh"
+#include "gc_clock.hh"
+#include "tombstone_gc.hh"
+#include "locator/token_metadata.hh"
+
+extern logging::logger dblog;
+
+class repair_history_map {
+public:
+    boost::icl::interval_map<dht::token, gc_clock::time_point, boost::icl::partial_absorber, std::less, boost::icl::inplace_max> map;
+};
+
+thread_local std::unordered_map<utils::UUID, seastar::lw_shared_ptr<repair_history_map>> repair_history_maps;
+
+static seastar::lw_shared_ptr<repair_history_map> get_or_create_repair_history_map_for_table(const utils::UUID& id) {
+    auto it = repair_history_maps.find(id);
+    if (it != repair_history_maps.end()) {
+        return it->second;
+    } else {
+        repair_history_maps[id] = seastar::make_lw_shared<repair_history_map>();
+        return repair_history_maps[id];
+    }
+}
+
+seastar::lw_shared_ptr<repair_history_map> get_repair_history_map_for_table(const utils::UUID& id) {
+    auto it = repair_history_maps.find(id);
+    if (it != repair_history_maps.end()) {
+        return it->second;
+    } else {
+        return {};
+    }
+}
+
+void drop_repair_history_map_for_table(const utils::UUID& id) {
+    repair_history_maps.erase(id);
+}
+
+get_gc_before_for_range_result get_gc_before_for_range(schema_ptr s, const dht::token_range& range, const gc_clock::time_point& query_time) {
+    bool knows_entire_range = true;
```

Comment:
I see this is something similar to what I suggested, but I don't understand how the caller can supply the range where it's the repair history table that supplies them. In any case please add some comments.

Reply:
It is used to get a gc_before for a sstable. The sstable provides the range. I will add comments here too.